core/src/AzureMcp.Core/Services/Azure/BaseAzureResourceService.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using Azure.ResourceManager.ResourceGraph;
+using Azure.ResourceManager.ResourceGraph.Models;
+using Azure.ResourceManager.Resources;
+using AzureMcp.Core.Options;
+using AzureMcp.Core.Services.Azure.Subscription;
+using AzureMcp.Core.Services.Azure.Tenant;
+using Microsoft.Extensions.Logging;
+
+namespace AzureMcp.Core.Services.Azure;
+
+/// <summary>
+/// Base class for Azure services that need to query Azure Resource Graph for resource management operations.
+/// Provides common methods for executing resource queries against Azure Resource Manager resources.
+/// </summary>
+public abstract class BaseAzureResourceService(
+    ISubscriptionService subscriptionService,
+    ITenantService tenantService,
+    ILoggerFactory? loggerFactory = null) : BaseAzureService(tenantService, loggerFactory)
+{
+    private readonly ISubscriptionService _subscriptionService = subscriptionService ?? throw new ArgumentNullException(nameof(subscriptionService));
+    private readonly ITenantService _tenantService = tenantService ?? throw new ArgumentNullException(nameof(tenantService));
+
+    /// <summary>
+    /// Cache for tenant resources by subscription tenant ID to avoid repeated enumeration of all tenants.
+    /// Key: Subscription's tenant ID, Value: TenantResource
+    /// </summary>
+    private readonly ConcurrentDictionary<Guid, TenantResource> _tenantCache = new();
+
+    /// <summary>
+    /// Gets the tenant resource for the specified subscription, using caching to avoid repeated tenant enumeration.
+    /// </summary>
+    /// <param name="tenantId">The tenant ID from the subscription</param>
+    /// <param name="cancellationToken">Cancellation token</param>
+    /// <returns>The tenant resource associated with the subscription</returns>
+    private async Task<TenantResource> GetTenantResourceAsync(Guid? tenantId, CancellationToken cancellationToken = default)
```

Comment:
TenantService should handle tenant caching.  See SubscriptionService for example

Reply:
I just realized the GetTenants already has caching, so remove the cache here